internal/federation/handle.go

```diff
@@ -128,6 +128,52 @@ func HandleMakeSendJoinRequests() func(*Server) {
 	}
 }
 
+// HandleInviteRequests is an option which makes the server process invite requests.
+//
+// inviteCallback is a callback function that if non-nil will be called and passed the incoming invite event
+func HandleInviteRequests(inviteCallback func(*gomatrixserverlib.Event)) func(*Server) {
+	return func(s *Server) {
+		// https://matrix.org/docs/spec/server_server/r0.1.4#put-matrix-federation-v2-invite-roomid-eventid
+		s.mux.Handle("/_matrix/federation/v2/invite/{roomID}/{eventID}", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			fedReq, errResp := gomatrixserverlib.VerifyHTTPRequest(
+				req, time.Now(), gomatrixserverlib.ServerName(s.ServerName), s.keyRing,
+			)
+			if fedReq == nil {
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			var inviteRequest gomatrixserverlib.InviteV2Request
+			if err := json.Unmarshal(fedReq.Content(), &inviteRequest); err != nil {
+				log.Printf(
+					"complement: Unable to unmarshal incoming /invite request: %s",
+					err.Error(),
+				)
+
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			if inviteCallback != nil {
+				inviteCallback(inviteRequest.Event())
+			}
+
+			// Send the response
+			res := map[string]interface{}{
+				"event": inviteRequest.Event(),
```

Comment:
Shouldn't we be signing this event or something?

Reply:
I guess so. The spec says we should, though synapse doesn't enforce it.

Reply:
... but it's a real pita to do so because gomatrixserverlib doesn't seem to expose the methods needed to do so?

Reply:
I think `inviteRequest.Event().Sign()` may work? https://github.com/matrix-org/gomatrixserverlib/blob/45159d0c62d0295dddeed9b951fd6ccbba1ca238/event.go#L558-L574